### PR TITLE
Add PyTorch Lightning Trainer for VQGAN

### LINKED ISSUE #13: Fix VQGAN to use PyTorch Lightning
Original VQGAN implementation used PyTorch Lightning module but we converted to PyTorch NN module. It is much easier to do distributed training with PyTorch Lightning, so let's convert it to Lightning module.

### LINKED ISSUE #14: Fix VQGAN notebook training loop
Since current VQGAN model is using NN module and not Lightning module, the training loop should handle calls such as optimizer.step, loss.backward(), optimizer.zero_grad(), etc.

### LINKED ISSUE #15: Add training script to fine-tune VQGAN
Using PyTorch Lightning, add an easy to use training/fine-tuning script.

### LINKED ISSUE #18: Update requirements.txt
`requirements.txt` is empty. Add required dependencies to it.

### PR DESCRIPTION
**Pull request by mahalrs**
Closes #13
Closes #14
Closes #15
Closes #18

This PR implements PyTorch Lightning Trainer for VQGAN. Tensorboard is used for logging. More details in the above mentioned issues.

### CODE REVIEW

**Comment by mahalrs** [comment]
@ArjunKrishna3367 @Waleedprw22

File: notebooks/vqgan.ipynb [hunk 1]
```diff
@@ -104,7 +91,12 @@
    },
    "outputs": [],
    "source": [
-    "model = VQModel.from_checkpoint('../src/checkpoints/vqgan/last.ckpt', config)"
+    "train_set = VQVisualNewsDataset('../src/data/visual_news_mini', 'train', transform)\n",
+    "train_loader = DataLoader(train_set,\n",
```
Comment (Waleedprw22) [inline notebooks/vqgan.ipynb:95]: The dataloader adjustment certainly makes sense here

**Comment by ArjunKrishna3367** [comment]
Couple of questions:
What is the purpose of process_pretrained_vqgan.py? It seems to just be loading the checkpoint and then saving it again.
What do the accelerator and strategy params do in the VQGAN training?

File: src/vqgan/lpips.py [hunk 2]
```diff
@@ -83,16 +83,16 @@ def __init__(self, chn_in, chn_out=1, use_dropout=False):
...
-    def __init__(self, requires_grad=False, pretrained=True):
+    def __init__(self, requires_grad=False, weights='DEFAULT'):
         super(vgg16, self).__init__()
-        vgg_pretrained_features = models.vgg16(pretrained=pretrained).features
-        self.slice1 = torch.nn.Sequential()
-        self.slice2 = torch.nn.Sequential()
-        self.slice3 = torch.nn.Sequential()
-        self.slice4 = torch.nn.Sequential()
-        self.slice5 = torch.nn.Sequential()
+        vgg_pretrained_features = models.vgg16(weights=weights).features
```
Comment (mahalrs) [inline src/vqgan/lpips.py:90]: `pretrained` is depreciated in the current version, and is replaced with `weights`.

**Comment by mahalrs** [comment]
> Couple of questions: What is the purpose of process_pretrained_vqgan.py? It seems to just be loading the checkpoint and then saving it again.

The pre-trained VQGAN checkpoint uses old PyTorch Lightning version. So, `python -m lightning.pytorch.utilities.upgrade_checkpoint vqgan.ckpt` is used to upgrade PyTorch Lightning module version.

Checkpoint also has `callbacks` that we need to modify, so `process_pretrained_vqgan.py` handles that. Also, for some reason, Lightning trainer is not able to resume training from the last epoch (maybe missing loops state_dict), so `process_pretrained_vqgan.py` is resetting that too.

**Comment by mahalrs** [comment]
> What do the accelerator and strategy params do in the VQGAN training?

`accelerator` and `strategy` are PyTorch Lightning parameters. `accelerator` is the device to use for training (e.g. cpu, gpu, tpu). `strategy` handles parallel processing such multi-gpu / multi-node training / multi-threaded / multi-process (e.g. ddp). You can read more here https://lightning.ai/docs/pytorch/latest/extensions/strategy.html

File: src/trainer_vqgan.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    train_set = VQVisualNewsDataset(args.dataset, 'train', transform)
+    train_loader = DataLoader(train_set,
+                              batch_size=args.train_batch,
+                              shuffle=True,
+                              num_workers=args.num_workers,
+                              pin_memory=True)
+
+    val_set = VQVisualNewsDataset(args.dataset, 'val', transform)
+    val_loader = DataLoader(val_set,
```
Comment (mahalrs) [inline src/trainer_vqgan.py:122]: Fix: should use DistributedSampler to ensure each process is using a different subset of the data. This will reduce chances of overfitting and improve model generation. https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler

File: src/trainer_vqgan.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        transforms.ToTensor(),
+        transforms.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5])
+    ])
+
+    # Load dataset
+    assert os.path.exists(args.dataset), f'{args.dataset} does not exist.'
+    assert os.path.isdir(args.dataset), f'{args.dataset} is not a directory.'
+
+    train_set = VQVisualNewsDataset(args.dataset, 'train', transform)
+    train_loader = DataLoader(train_set,
```
Comment (mahalrs) [inline src/trainer_vqgan.py:115]: Fix: should use DistributedSampler to ensure each process is using a different subset of the data.

File: src/trainer_vqgan.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+    val_set = VQVisualNewsDataset(args.dataset, 'val', transform)
+    val_loader = DataLoader(val_set,
+                            batch_size=args.val_batch,
+                            shuffle=False,
+                            num_workers=args.num_workers,
+                            pin_memory=True)
+
+    test_set = VQVisualNewsDataset(args.dataset, 'test', transform)
+    test_loader = DataLoader(test_set,
```
Comment (mahalrs) [inline src/trainer_vqgan.py:129]: Fix: should use DistributedSampler to ensure each process is using a different subset of the data.

**Comment by Waleedprw22** [comment]
We should probably mention that it is very easy for these large checkpoint files to get corrupted and so if you run into a pickling error, may want to delete and re-upload. It is best to use these to smoothly and quickly download these files:

.... LINKS REMOVED ....

PS: I should delete this comment since previous edits are still visible but for now I think we are good.

**Review by Waleedprw22 — APPROVED** [review]
The tests were successful and smooth. I think we just need to mention in the README or somewhere to use these commands to download the checkpoints:

.... LINKS REMOVED ....

PS: I should delete this comment since previous edits are still visible but for now I think we are good.

**Comment by mahalrs** [comment]
> We should probably mention that it is very easy for these large checkpoint files to get corrupted and so if you run into a pickling error, may want to delete and re-upload. It is best to use these to smoothly and quickly download these files:
> ...

This comment has links to our cloud storage bucket, so I'm going to delete it. We shouldn't post links to public buckets here. Also, we are not planning to keep these buckets forever ($$$) so we shouldn't mention them but we can add a note about pickling error/corrupted files.

File: src/vqgan/lpips.py (hunk 2 above)
Comment (mahalrs) [reply src/vqgan/lpips.py:90]: /resolved

File: src/trainer_vqgan.py (hunk 3 above)
Comment (mahalrs) [reply src/trainer_vqgan.py:122]: /resolved

File: src/trainer_vqgan.py (hunk 4 above)
Comment (mahalrs) [reply src/trainer_vqgan.py:115]: /resolved

File: src/trainer_vqgan.py (hunk 5 above)
Comment (mahalrs) [reply src/trainer_vqgan.py:129]: /resolved

**Review by ArjunKrishna3367 — APPROVED** [review]
Everything ran fine for me as well, looks good! No outstanding questions/concerns.

**Review by mahalrs — COMMENTED** [review]
Few changes requested
